DRAFT: Standard template for On-prem deployment  for  public use

ewnetu opened the pull request:
This is a draft documentation regarding generic documentation about on-prem CK8S deployment.  I am sharing this draft to see if this is the right direction and get  input before diving into details.

cristiklein commented on docs/operator-manual/on-prem-standard.md:
```diff
@@ -0,0 +1,276 @@
+
+## Prerequisites
+
+https://elastisys.io/compliantkubernetes/operator-manual/capacity-management/
```
I'm unsure to understand why this line is present.

Also, I mirroring our internal docs, I would start with:

```
1. Check the infra-provider. If in doubt, do a [provider audit](link to audit page).
```

ewnetu replied:
That was by mistake :)

cristiklein commented on docs/operator-manual/on-prem-standard.md:
```diff
@@ -0,0 +1,276 @@
+
+    > **_NOTE:_** the following steps are done from the root of the ck8s-ops repository
+
+    > **_WARNING:_** When using the devbox, your local ck8s-ops repository is used for templates, so make sure it is up to date
```
This line can be removed.

cristiklein commented on docs/operator-manual/on-prem-standard.md:
````diff
@@ -0,0 +1,276 @@
+1. Clone the Elastisys Compliant Kubernetes Kubespray repo as follows:
+
+    ```console
+    git clone --recursive https://github.com/elastisys/compliantkubernetes-kubespray
````
Would we clone in the config repo? In that case, we should use `git submodule` here.

ewnetu replied:
I think we shouldn't clone it in the config repo. I checked how the devops people work and they don't put the ck8s (kubespray and apps) repos in their environments config repo. I think we should follow that and allow the user to clone these repos as they prefer.

cristiklein commented on docs/operator-manual/on-prem-standard.md:
````diff
@@ -0,0 +1,276 @@
+1. Clone compliantkubernetes-apps and Install Pre-requisites as follows:
+
+    ```console
+    https://github.com/elastisys/compliantkubernetes-apps.git
````
`git clone` I assume?

Same question as above. Do we clone outside the config repo (and then we need to specify where) or do we clone inside the config repo (and then we need to use `git submodule`). What do you think works best?

cristiklein commented on docs/operator-manual/on-prem-standard.md:
````diff
@@ -0,0 +1,276 @@
+### Init Kubespray config in your config path.
+
+    ```console
+    export CK8S_KUBESPRAY_REPOSITORY_PATH=/path/to/compliantkubernetes-kubespray
````
We should already set the correct path, once we decided where to clone the Kubespray layer.

cristiklein commented on docs/operator-manual/on-prem-standard.md:
````diff
@@ -0,0 +1,276 @@
+    done
+    ```
+
+For Citycloud and ElastX environments, edit the `ck8s-k8s-cluster-openstack.yaml` file and set the following variables:
````
```suggestion
For OpenStack-based cloud providers, edit the `ck8s-k8s-cluster-openstack.yaml` file and set the following variables:
```

cristiklein commented on docs/operator-manual/on-prem-standard.md:
```diff
@@ -0,0 +1,276 @@
+
+### Configure OIDC
+
+To configure OIDC access for kubernetes api and  other services dex should be configured with your identity provider. Check what dex needs from [your identity provider](https://dexidp.io/docs/connectors/).
```
Nice! Minor capitalization fixes:

```suggestion
To configure OpenID access for Kubernetes API and other services, Dex should be configured with your identity provider. Check what Dex needs from [your identity provider](https://dexidp.io/docs/connectors/).
```

cristiklein commented on docs/operator-manual/on-prem-standard.md:
```diff
@@ -0,0 +1,276 @@
+
+#### Configure OIDC endpoint
+
+Set `kube_oidc_url` in `${CK8S_CONFIG_PATH}/sc-config/group_vars/k8s_cluster/ck8s-k8s-cluster.yaml` and `${CK8S_CONFIG_PATH}/wc-config/group_vars/k8s_cluster/ck8s-k8s-cluster.yaml` based on your cluster. For example, if your domain is `example.com` then kube_oidc_url is set as `kube_oidc_url: https://dex.example.com` in the both files.
```
```suggestion
Set `kube_oidc_url` in `${CK8S_CONFIG_PATH}/sc-config/group_vars/k8s_cluster/ck8s-k8s-cluster.yaml` and `${CK8S_CONFIG_PATH}/wc-config/group_vars/k8s_cluster/ck8s-k8s-cluster.yaml` based on your cluster. For example, if your domain is `example.com` then kube_oidc_url is set as `kube_oidc_url: https://dex.example.com` in both files.
```

cristiklein commented on docs/operator-manual/on-prem-standard.md:
```diff
@@ -0,0 +1,276 @@
+
+### Copy the VMs information to the inventery files
+
+ Add the host name,  user and ip address of each VM that you prepared above in `${CK8S_CONFIG_PATH}/sc-config/inventory.ini`for service cluster and `${CK8S_CONFIG_PATH}/sc-config/inventory.ini` for workload cluster. Moreover, you also need to add the host names of the master nodes under `[kube_control_plane]`, etdc nodes under `[etcd]` and worker nodes under `[kube_node]`.  
```
Isn't the user always going to be `ubuntu`?

Also, aren't we always going to use the same Nodes for `etcd` as for `kube_control_plane`?

```suggestion
 Add the host name, user and IP address of each VM that you prepared above in `${CK8S_CONFIG_PATH}/sc-config/inventory.ini`for service cluster and `${CK8S_CONFIG_PATH}/sc-config/inventory.ini` for workload cluster. Moreover, you also need to add the host names of the master nodes under `[kube_control_plane]`, etdc nodes under `[etcd]` and worker nodes under `[kube_node]`.  
```

cristiklein commented on docs/operator-manual/on-prem-standard.md:
```diff
@@ -0,0 +1,276 @@
+
+ Add the host name,  user and ip address of each VM that you prepared above in `${CK8S_CONFIG_PATH}/sc-config/inventory.ini`for service cluster and `${CK8S_CONFIG_PATH}/sc-config/inventory.ini` for workload cluster. Moreover, you also need to add the host names of the master nodes under `[kube_control_plane]`, etdc nodes under `[etcd]` and worker nodes under `[kube_node]`.  
+  
+  > **_NOTE:_** Make sure that the user has ssh access to the VMs.
```
```suggestion
  > **_NOTE:_** Make sure that the user has SSH access to the VMs.
```

cristiklein commented on docs/operator-manual/on-prem-standard.md:
````diff
@@ -0,0 +1,276 @@
+### Initialize the apps configuration
+
+```bash
+export  CK8S_APPS_REPOSITORY_PATH=</path/to/compliantkubernetes-kubespray-apps>
````
Once we decided above where to clone the apps layer, we can already set the right path here.

cristiklein commented on docs/operator-manual/on-prem-standard.md:
````diff
@@ -0,0 +1,276 @@
+```console
+$CK8S_APPS_REPOSITORY_PATH/compliantkubernetes-apps/bin/ck8s apply sc
+$CK8S_APPS_REPOSITORY_PATH/compliantkubernetes-apps/bin/ck8s apply wc
+done
````
I guess we need to remove `done`.

cristiklein reviewed (changes requested):
Nice! I feel this is a huge step in the right direction.

There is still some work left to be done, but overall great work!

Food for thought: Should we have a `> **DECISION-TO-BE-TAKEN**` tag, which we should replace with the taken decision when implementing such projects?

I see the following decisions:

- Overall architecture, i.e., VM sizes, load-balancer configuration, storage configuration, etc.
- Identity Provider (IdP) choice and configuration
- On-call Management Tool (OMT) choice and configuration

Other decisions, such as changes to `sc-config.yaml` and `wc-config.yaml` could be documented in the `git commit` message or as comment in the file itself.

ewnetu commented on docs/operator-manual/on-prem-standard.md:
```diff
@@ -0,0 +1,276 @@
+
+https://elastisys.io/compliantkubernetes/operator-manual/capacity-management/
+1. Prepare ubuntu-based VMs 
+    If you are using public clouds, you can use [Kubespray](https://github.com/kubernetes-sigs/kubespray) to create the VMs.
```
Azure  uses [AzureRM(resource manager)](https://github.com/kubernetes-sigs/kubespray/tree/master/contrib/azurerm) instead of Terraform. Maybe that is an exception.

cristiklein replied:
Sounds good. Please document it accordingly. Perhaps we can have a bullet list:

If you are using public clouds, you can use the scripts included in Kubespray:

* For Azure, use [AzureRM scripts](link).
* For other clouds, use their respective [Terraform scripts](link).

moule3053 replied:
I think it would be nice to use Terraform as much as possible. I see that Terraform also has an Azure Provider https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs

cristiklein replied:
@moule3053 I agree that it would be great to use Terraform with Azure too. However, Kubespray currently provides no script for that, so I'm worried that: (a) writing one would be quite some undertaking, (b) we wouldn't be able to maintain it.

moule3053 commented on docs/operator-manual/on-prem-standard.md:
```diff
@@ -0,0 +1,276 @@
+# Standard Template for on-prem Environment
+
+This document contains the instructions on how to setup a new Compliant Kubernetes on-prem environment.
```
```suggestion
This document contains instructions on how to set up a new Compliant Kubernetes on-prem environment.
```

moule3053 commented on docs/operator-manual/on-prem-standard.md:
````diff
@@ -0,0 +1,276 @@
+    ```
+
+1. Create the domain name. 
+    You need to create a domain name to access the different services in your environment. You will need to set up the following DNS entries (replace example.com with your domain).
````
```suggestion
    You need to create a domain name to access the different services in your environment. You will need to set up the following DNS entries (replace `example.com` with your domain name).
```

moule3053 commented on docs/operator-manual/on-prem-standard.md:
```diff
@@ -0,0 +1,276 @@
+
+1. Create S3 credentials and add them to `.state/s3cfg.ini`.
+
+1. Set up Loadbalancer
+
+You need to setup two loadbalancers, one for workload cluster and one for service cluster.
```
```suggestion
1. Set up load balancers

You need to set up two load balancers, one for the workload cluster and one for the service cluster.
```

moule3053 commented on docs/operator-manual/on-prem-standard.md:
````diff
@@ -0,0 +1,276 @@
+    git clone --recursive https://github.com/elastisys/compliantkubernetes-kubespray
+    done
+    ```
+1. Clone compliantkubernetes-apps and Install Pre-requisites as follows:
````
```suggestion
1. Clone compliantkubernetes-apps and install pre-requisites as follows:
```

moule3053 commented on docs/operator-manual/on-prem-standard.md:
```diff
@@ -0,0 +1,276 @@
+  
+  > **_NOTE:_** Make sure that the user has ssh access to the VMs.
+
+### Run kubespray to deploy the Kubernetes clusters
```
```suggestion
### Run Kubespray to deploy the Kubernetes clusters
```

moule3053 commented on docs/operator-manual/on-prem-standard.md:
````diff
@@ -0,0 +1,276 @@
+ done
+```
+
+Please restart the operator pod, `rook-ceph-operator*`, if some pods stalls in initialization state as shown below:
````
```suggestion
Please restart the operator pod, `rook-ceph-operator*`, if some pods stall in the initialization state as shown below:
```

moule3053 commented on docs/operator-manual/on-prem-standard.md:
````diff
@@ -0,0 +1,276 @@
+done
+```
+
+You should see PVCs in Bound state. If you want to clean the previously created PVCs:
````
```suggestion
You should see PVCs in `Bound` state. If you want to clean the previously created PVCs:
```

moule3053 commented on docs/operator-manual/on-prem-standard.md:
````diff
@@ -0,0 +1,276 @@
+
+vim ${CK8S_CONFIG_PATH}/common-config.yaml
+```
+Edit the secrets.yaml file and add the credentials for:
+- s3 - used for backup storage
+- dex - connectors -- check [your indentiy provider](https://dexidp.io/docs/connectors/).
````
Make this multi-line?

cristiklein replied:
Please also add the On-Call Management Tool here.

moule3053 commented on docs/operator-manual/on-prem-standard.md:
```diff
@@ -0,0 +1,276 @@
+
+### Create S3 buckets
+
+You can use the following command to create required S3 buckets. The command uses s3cmd in the background and gets configuration and credentials for your S3 provider from the `~/.s3cfg` file.
```
```suggestion
You can use the following command to create the required S3 buckets. The command uses `s3cmd` in the background and gets configuration and credentials for your S3 provider from the `~/.s3cfg` file.
```

moule3053 commented on docs/operator-manual/on-prem-standard.md:
````diff
@@ -0,0 +1,276 @@
+done
+```
+
+Check the output of the command above. All Pods needs to be Running or Completed.
````
```suggestion
Check the output of the command above. All Pods need to be `Running` or `Completed` status.
```

moule3053 commented on docs/operator-manual/on-prem-standard.md:
````diff
@@ -0,0 +1,276 @@
+```
+
+Check the output of the command above.
+All resources need to have the Ready column True.
````
```suggestion
All resources need to have the `Ready` column `True`.
```

moule3053 reviewed (changes requested):
Mostly made suggestions to improve style, i.e. capitalization, key works, plurals, etc. for now.

cristiklein commented on docs/operator-manual/on-prem-standard.md:
```diff
@@ -0,0 +1,250 @@
+
+1. Create the domain name.
+    You need to create a domain name to access the different services in your environment. You will need to set up the following DNS entries (replace example.com with your domain name).
+    - Point these domains to the workload cluster ingress controller:
```
:thinking: The Ingress Controller doesn't really exist at this point, but I do understand the need to "do away" with DNS early in the process. Any idea how we can solve this chicken-and-egg problem?

ewnetu replied:
@cristiklein I have already addressed @moule3053's feedback :)

cristiklein replied:
Regarding addressing @moule3053 's feedback, I'm confused. Here is what I see:
![Screenshot from 2022-10-14 12-08-09](https://user-images.githubusercontent.com/1660833/195822305-99208176-778d-4dba-875d-8f5fa1391a17.png)

Regarding supported on-call tools, we currently only support Slack and OpsGenie (see config [here](https://github.com/elastisys/compliantkubernetes-apps/blob/main/config/config/sc-config.yaml#L992)). With code changes, we could support [all of these](https://prometheus.io/docs/alerting/latest/configuration/).

cristiklein reviewed (changes requested):
Nice @ewnetu !

Can you please resolve the remaining comments from me and @moule3053 .

Afterwards, I'd say that this is as good as it gets internally, and the next project will be the real-life test.

ewnetu commented:
> Nice @ewnetu !
> 
> Can you please resolve the remaining comments from me and @moule3053 .
> 
> Afterwards, I'd say that this is as good as it gets internally, and the next project will be the real-life test.

@cristiklein  I have addressed the the feedback. Let me know if there is anything  left.

cristiklein reviewed (approved):
Awesome! I feel this is as far as we can bring it internally. The next on-prem project will be the true test.
